### DISCUSSION TITLE
GRAFF tutorial

### DISCUSSION THREAD

**Discussion by realfolkcode** [pyg-team/pytorch_geometric#7268] [Show and tell]
In [Graph Neural Networks as Gradient Flows: understanding graph convolutions via energy](https://arxiv.org/pdf/2206.10991.pdf), the authors propose to parametrize the energy functional, and differentiate the energy to obtain GNNs that work well in both heterophilic and homophilic settings. The paper is amazing but I couldn't find an implementation anywhere. So I decided to write it by myself. I made a tutorial where I briefly cover theoretical aspects of GRAFF, and implement a diagonally-dominant variant of GRAFF based on GCN. The tutorial also covers how to parametrize the weights of layers based on PyTorch's functionality. The results of the paper are informally recreated on `WebKB` and `Cora` datasets.

I would be greatful if you could review my tutorial and implementation, as I am not 100% sure it is correct (although it works well).

[Link to the tutorial](https://github.com/realfolkcode/GRAFF/blob/main/GRAFF_Tutorial_PyG.ipynb)

**Comment by rusty1s** [comment]
Thanks for your efforts, this looks cool :) Do you want to have this tutorial listed in the "External Resources" page in the documentation?

**Reply by realfolkcode** [reply]
Sure! It would be a great honor for me!
